gadget/mountedfilesystem.go

```diff
@@ -922,17 +896,24 @@ func (f *mountedFilesystemUpdater) rollbackFile(dstRoot, source, target string,
 			return err
 		}
 	} else {
-		// none of the markers exists, file is not preserved, meaning, it has
-		// been added by the update
-		if err := os.Remove(dstPath); err != nil && !os.IsNotExist(err) {
-			return fmt.Errorf("cannot remove written update: %v", err)
+		// no backup marker, could be a new file or one that got
+		// preserved by request of the observer, in which case there
+		// will be a stamp
+		if !osutil.FileExists(preserveStamp) {
+			if err := os.Remove(dstPath); err != nil && !os.IsNotExist(err) {
+				return fmt.Errorf("cannot remove written update: %v", err)
+			}
+			// since it's a new file, there was no original content
+			data.Before = ""
+		} else {
+			// content was preserved at original location
+			data.Before = dstPath
```

Comment:
the observe below will be a no-op anyway in this case? should we just return here?